refactor(popup): migrate to fast element

rinaok opened the pull request:
`vwc-elevation` and `vwc-icon-button` need to be migrated to accomplish this PR.

rinaok commented:
@yinonov @YonatanKra  I receive the following error when calling `computePosition`: 
`Uncaught (in promise) ReferenceError: process is not defined`.

<img width="500" alt="Screen Shot 2022-03-02 at 18 25 44" src="https://user-images.githubusercontent.com/10883919/156403938-717151d2-675a-4ca0-a9a1-6b8e8b3ed964.png">
<img width="692" alt="Screen Shot 2022-03-02 at 18 26 06" src="https://user-images.githubusercontent.com/10883919/156403994-64669d3b-a869-4013-942d-b0ed6dcc2a22.png">

Found this issue: https://github.com/floating-ui/floating-ui/issues/1506

YonatanKra commented:
@rinaok - I don't mean the version in `package.json`.
You should use `@floating-ui/dom` and not `core` IMO.

yinonov commented on libs/components/src/lib/popup/popup.ts:
```diff
@@ -138,11 +138,10 @@ export class Popup extends FoundationElement {
 			return;
 		}
-		if (!this.popupEl) {
+		if (!this.popupEl || this.popupEl === undefined) {
```
same here

yinonov reviewed (commented):
if popup strategy is set to _fixed_, it should probably close the popup upon scroll. similar to the browsers strategy. WDYT?

rinaok commented on libs/components/src/lib/popup/popup.spec.ts:
```diff
@@ -0,0 +1,77 @@
+
+const COMPONENT_TAG = 'vwc-popup';
+
+global.ResizeObserver = jest.fn().mockImplementation(() => ({
+	observe: jest.fn(),
+	unobserve: jest.fn(),
+	disconnect: jest.fn(),
+}));
```
@YonatanKra  @yinonov 
Jest fails with an error message stating that ResizeObserver inside floating-ui classes is undefined.
That fixed the problem.

rinaok commented:
> if popup strategy is set to _fixed_, it should probably close the popup upon scroll. similar to the browsers strategy. WDYT?

yes, I agree. It does look strange. I've added `hide()` Middleware and removed the `shift() `Middleware. I think it's better now. Tell me what you think.

rachelbt commented on libs/components/src/lib/popup/popup.scss:
```diff
@@ -0,0 +1,43 @@
+@use 'sass:map';
+@use '../../../../shared/src/lib/sass/mixins/typography' as typography;
```
not sure what we need this for

rinaok replied:
removed

YonatanKra reviewed (changes requested):
A really small change :)

rachelbt commented on libs/components/src/lib/popup/popup.scss:
```diff
@@ -0,0 +1,43 @@
+	--border-radius-size: 6px;
+
+	width: min-content;
+	background-color: var(#{--vvd-color-surface-4dp});
```
there's non need for `#{} `if you are not using a scss variable here.
But we should use them :)
you can set a variable in the start of the file (if we would have more than one its better to open.a new file for it) and set:
`$popup-bg: --vvd-color-surface-4dp;
`and use it in the CSS here and in the arrow bg like this:
`background-color: var(#{$popup-bg});`

rinaok replied:
done :)

yinonov commented on change/@vonage-vivid-fff24903-c6b1-40dd-9364-7f55a550515e.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "major",
```
Popup is alpha. And is probably underlying component for now. I vote 'none'

YonatanKra commented:
Note that the coverage is very small (reduces our whole repo's coverage by 5%).
![image](https://user-images.githubusercontent.com/6459899/158057288-b471a26f-89d7-4fa6-a2fc-bff61553c75c.png)
Seems like some features are not tested:
1. `aria-hidden` is never set to false in the tests.
2. `part=vvid-theme-alternate` is a not-tested branch.
3. `renderArrow` is not called during the tests.
4. `renderDismissButton` is not called during the tests.

rinaok commented on libs/components/src/lib/popup/README.md:
```diff
@@ -0,0 +1,176 @@
+Use the `strategy` attribute to set the position of the popup.
+
+- Type: `'fixed'` | `'absolute'`
+- Default: `'fixed'`
```
https://floating-ui.com/docs/computePosition#strategy

rachelbt replied:
according to floating UI our default is absolute...

rinaok replied:
@rachelbt It is absolute by their default, yes. But our default is fixed.
@yinonov Why we changed the default to fixed?

yinonov commented:
> > if popup strategy is set to _fixed_, it should probably close the popup upon scroll. similar to the browsers strategy. WDYT?
> 
> @yinonov yes, I agree. It does look strange. I've added `hide()` Middleware and removed the `shift() `Middleware. I think it's better now. Tell me what you think.

Maybe we should add html snippet for fixed and express this exact dialog to users?

yinonov commented on libs/components/src/lib/popup/README.md:
````diff
@@ -0,0 +1,178 @@
+</div>
+
+```
+## Z-index
````
```suggestion
## Top-layer
```

rinaok commented:
> > > if popup strategy is set to _fixed_, it should probably close the popup upon scroll. similar to the browsers strategy. WDYT?
> > 
> > 
> > @yinonov yes, I agree. It does look strange. I've added `hide()` Middleware and removed the `shift() `Middleware. I think it's better now. Tell me what you think.
> 
> Maybe we should add html snippet for fixed and express this exact dialog to users?

@yinonov We already have a snippet for strategy, what do you want to show there?

YonatanKra commented on libs/components/src/lib/popup/popup.spec.ts:
```diff
@@ -0,0 +1,110 @@
+		});
+	});
+
+	describe('render classes', () => {
+		it('should render arrow class', async () => {
+			element.arrow = true;
+			await elementUpdated(element);
+			expect(element.shadowRoot?.querySelector('.arrow')).not.toBeNull();
+		});
+
+		it('should render dismiss class', async () => {
+			element.dismissible = true;
+			await elementUpdated(element);
+			expect(element.shadowRoot?.querySelector('.dismissible')).not.toBeNull();
+		});
+	});
```
These are two different API's.
```suggestion
	describe('render arrow', () => {
		it('should render an arrow icon if arrow is true', async () => {
			// TODO
		});
		
		it('should not render an arrow icon if arrow is false', async () => {
			// TODO
		});
		
		it('should remove the arrow class on the container if arrow is false', () => {
		        const controlElement = element.shadowRoot?.querySelector('.control');
			expect(controlElement.classList.contains('.arrow')).toEqual(false);
		});
		
		it('should set the arrow class on the container if arrow is true', () => {
		        element.arrow = true;
			await elementUpdated(element);
			const controlElement = element.shadowRoot?.querySelector('.control');
			expect(controlElement.classList.contains('.arrow')).toEqual(true);
		});
	});
```
This suggestion only covers the idea for `arrow`. Do the same for dismissible.

rinaok replied:
@YonatanKra It won't work, the arrow and dismiss classes are not inside the control.
Also the "should remove the arrow class on the container if arrow is false" check is not necessary cause "initializes as a vwc-popup" test already checks it.

YonatanKra replied:
That's exactly why we should add the correct test to document where we should expect the class.
BTW - is this class exposed as API? I think what's more important is that we have an icon/button, no?

YonatanKra commented on libs/components/src/lib/popup/popup.spec.ts:
```diff
@@ -0,0 +1,110 @@
+	});
+
+	describe('accessibility', () => {
+		it('should set aroa-hidden', async () => {
```
```suggestion
		it('should set aria-hidden', async () => {
```

yinonov commented:
https://github.dev/Vonage/vivid-3/blob/374ff2f9087b8ff73e0a60b3d54067c1cf1f4225/libs/components/rollup.config.prod.ts#L41-L44

yinonov commented on libs/components/rollup.config.prod.ts:
```diff
@@ -36,7 +38,10 @@ module.exports = function setVividRollupConfig(config) {
 
 
-	const plugins = [...config.plugins];
+	const plugins = [...config.plugins,
+		replace({
+			'process.env.NODE_ENV': JSON.stringify(process.env.NODE_ENV),
+		})];
```
This should be checked thoroughly

YonatanKra replied:
That's the official solution

yinonov commented on libs/components/src/lib/popup/popup.ts:
```diff
@@ -0,0 +1,184 @@
+	 */
+	@attr anchor: string = '';
+
+	/**
+	 * popup's anchor element
+	 *
+	 * @private
+	 */
+	@attr private anchorEl: Element | null | undefined;
```
why is this an interface? wasn't it a getter? have we decided to expose it as well?
we already have `anchor` member. it can, alone, accept _string_ or _HTMLElement_

rinaok replied:
it's private, it's not an interface

yinonov commented on libs/components/src/lib/popup/popup.spec.ts:
```diff
@@ -77,10 +77,10 @@ describe('vwc-popup', () => {
+		it('should set the dismiss class on the container if dismissible is true', async () => {
 			element.dismissible = true;
 			await elementUpdated(element);
 			expect(element.shadowRoot?.querySelector('.dismissible')).not.toBeNull();
```
I know we talked about this today but I missed the part where `getControlElement(element)` is actually responsible for querying the `.control`. so we better use it

yinonov commented on change/@vonage-vivid-fff24903-c6b1-40dd-9364-7f55a550515e.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "minor",
```
exist in 2
```suggestion
  "type": "none",
```

yinonov commented on libs/components/src/lib/enums.ts:
```diff
@@ -35,14 +35,6 @@ export enum Size {
 	BaseLarge = 'base-large',
 }
 
-export enum Position {
-	Top = 'TOP',
-	Bottom = 'BOTTOM',
-	Start = 'START',
-	Center = 'CENTER',
-	End = 'END',
-}
```
where will this move to? we might wanna have a single source of truth for these types. components using popup under the hood might need to proxy it

rinaok replied:
I didn't move it. I'm using floating-ui's enums.

yinonov commented on libs/components/src/lib/popup/README.md:
````diff
@@ -0,0 +1,211 @@
+  </vwc-popup>
+</div>
+```
+## Top-layer
````
```suggestion
## Caveat
```

yinonov commented on libs/components/src/lib/popup/popup.template.ts:
```diff
@@ -0,0 +1,46 @@
+	</div>`;
+
+const renderDismissButton = () => {
+	return html`<vwc-button @click="${x => x.handleDismissClick()}"
```
small maybe?

```suggestion
	return html`<vwc-button size="base-small" @click="${x => x.handleDismissClick()}"
```

yinonov commented on libs/components/src/lib/popup/ui.test.ts:
```diff
@@ -0,0 +1,52 @@
+import * as path from 'path';
```
snapshots are messed up. like, not anchored properly

yinonov commented on libs/components/src/lib/popup/ui.test.ts:
```diff
@@ -0,0 +1,52 @@
+
+	await loadComponent({
+		page,
+		componentName,
```
how did you import more components to the page? I'm seeing _icon_ which isn't imported by _popup_

yinonov replied:
nice! 🥇

YonatanKra commented on libs/components/src/lib/popup/README.md:
```diff
@@ -0,0 +1,211 @@
+
+Use the `anchor` attribute to reference the ID to element in the popup’s owner document.
+
+Since the popup is positioned absolutely to its anchor, it is best to wrap it with a div element and set its position `relative`.
```
```suggestion
The popup is positioned absolutely to its anchor.  It is recommended to wrap it with a div element and set its position `relative`.
```
Would also appreciate a short description like: "how things can go wrong" or "what adding the wrapper solves" sentence.

YonatanKra commented on libs/components/src/lib/popup/README.md:
````diff
@@ -0,0 +1,211 @@
+</script>
+```
+## Arrow
+Use the `arrow` attribute to add small triangle to indicate the trigger element.
````
```suggestion
Adds a small triangle to indicate the trigger element.
```

YonatanKra commented on libs/components/src/lib/popup/README.md:
```diff
@@ -0,0 +1,211 @@
+
+## Dismissible
+
+Use the `dismissible` attribute to add close button to the popup.
```
```suggestion
Adds a close button to the popup.
```

YonatanKra commented on libs/components/src/lib/popup/README.md:
````diff
@@ -0,0 +1,211 @@
+</div>
+```
+## Alternate
+Use the `alternate` attribute to set the color-scheme to dark.
````
```suggestion
Sets the color-scheme to dark or light (depending on current user's system settings).
```

YonatanKra commented on libs/components/src/lib/popup/README.md:
````diff
@@ -0,0 +1,211 @@
+```
+## Corner
+
+Use the `corner` attribute to set the placement of the popup.
````
```suggestion
Sets the placement of the popup around the anchor.
```
Is this correct? Around the anchor?

rinaok replied:
yes

YonatanKra commented on libs/components/src/lib/popup/popup.ts:
```diff
@@ -0,0 +1,187 @@
+		this.corner = 'left'; // default corner
+	}
+
+	override connectedCallback(): void {
+		super.connectedCallback();
+	}
```
What's that for?

YonatanKra commented on libs/components/src/lib/popup/popup.ts:
```diff
@@ -0,0 +1,187 @@
+
+	override disconnectedCallback(): void {
+		super.disconnectedCallback();
+		if (this.cleanup) { this.cleanup(); }
```
Is `cleanup` an api? It is not documented and not tested. Why would cleanup no exist?

rinaok replied:
You can read about cleanup here: https://floating-ui.com/docs/autoUpdate.

YonatanKra commented on libs/components/src/lib/popup/popup.ts:
```diff
@@ -0,0 +1,187 @@
+		Object.assign(this.popupEl.style, {
+			left: `${popupX}px`,
+			top: `${popupY}px`,
+			visibility: referenceHidden ? 'hidden' : 'visible',
```
One of these cases is not tested. Don't know which without running locally.

YonatanKra commented on libs/components/src/lib/popup/popup.ts:
```diff
@@ -0,0 +1,187 @@
+
+	private assignArrowPosition(data: any): void {
+		const { x: arrowX, y: arrowY } = data.middlewareData.arrow;
+		const staticSide: any = { top: 'bottom', right: 'left', bottom: 'top', left: 'right' };
```
Why not set as constant at the top of the file?
And about the content - it is a WTF moment :)
Can you explain the usecase plz? There's no test for it. The function `updatePosition` is an open API (public) but not tested, so hard to tell what this function is supposed to do.

YonatanKra commented on libs/components/src/lib/components.ts:
```diff
@@ -7,5 +7,5 @@ export * from './layout';
 export * from './elevation';
 export * from './breadcrumb-item';
 export * from './breadcrumb';
-export * from './side-drawer';
```
What happened to side-drawer?

YonatanKra commented on libs/components/src/lib/popup/ui.test.ts:
```diff
@@ -0,0 +1,30 @@
+
+const components = ['popup', 'text', 'icon', 'button'];
+test('should show the component', async ({ page }: { page: Page }) => {
+	const template = extractHTMLBlocksFromReadme(path.join(__dirname, 'README.md'))
```
Are these use cases include all of the `top`/`bottom`/`left`/`right` use cases possible?

YonatanKra replied:
I think in this case, because we have so many use cases for positioning, it is worthwhile to hard code them here instead of taking out of docs an incomplete list of them.

YonatanKra commented on change/@vonage-vivid-fff24903-c6b1-40dd-9364-7f55a550515e.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "none",
```
Why is the type `none`? Why not `minor`? Isn't it a new feature?

rinaok replied:
@yinonov changed it to none

YonatanKra commented on libs/shared/src/lib/test-utils/index.spec.ts:
```diff
@@ -29,6 +29,11 @@ describe(`test-utils`, function () {
       expect(element.id).toEqual('test');
       expect(element.parentElement?.parentElement?.tagName).toEqual('BODY');
     });
+    it(`should return a dom element under div under body with add`, function () {
```
![image](https://user-images.githubusercontent.com/6459899/161707341-b8853ea3-7ae6-4c17-9bf7-5df047873623.png)

YonatanKra commented on libs/components/src/lib/popup/popup.spec.ts:
```diff
@@ -0,0 +1,262 @@
+			element.anchor = 'anchor';
+			await elementUpdated(element);
+			element.disconnectedCallback();
+			expect(true).toBeTruthy();
```
This is not the test that should be here.
You should `spyOn` the `element.updatePosition` method and see that it stops running after the element disconnects if you change a position.

YonatanKra commented on libs/components/src/lib/popup/popup.spec.ts:
```diff
@@ -0,0 +1,262 @@
+describe('vwc-popup', () => {
+	let element: Popup;
+
+	global.ResizeObserver = jest.fn()
```
I think this should be defined in a `beforeEach` - otherwise `element.updatePosition` might differ in every test. Also, for testing the cleanup, you should set `observe` to run a spy.
`observeSpy = jest.fn(element.updatePosition)`
and then:
`observe: observeSpy,`

YonatanKra commented on libs/components/src/lib/popup/popup.ts:
```diff
@@ -0,0 +1,189 @@
+		this.open = false;
+	}
+
+	handleDismissClick(): void {
```
Is this tested? Do we actually test the `click` event?

YonatanKra commented on libs/components/src/lib/popup/popup.spec.ts:
```diff
@@ -0,0 +1,273 @@
+
+			jest.clearAllMocks();
+			element.disconnectedCallback();
+			element.anchor = '';
```
This is unclear. We need to change something that triggers the observer.

YonatanKra commented on libs/components/src/lib/popup/popup.ts:
```diff
@@ -0,0 +1,185 @@
+			}
+		}
+		if (this.anchorEl && this.popupEl) {
+			this.cleanup = autoUpdate(this.anchorEl, this.popupEl, () => this.updatePosition());
```
now I see something else that's troubling - shouldn't we run `clean` if `anchor` is undefined?
I mean - if we remove the `anchor` from the DOM, the observer is still there which is a possible memory leak.

rinaok replied:
yes, you are right

YonatanKra commented on libs/components/src/lib/popup/ui.test.ts:
```diff
@@ -0,0 +1,105 @@
+} from '../../visual-tests/visual-tests-utils';
+
+const components = ['popup', 'text', 'icon', 'button'];
+test('should show the component', async ({ page }: { page: Page }) => {
```
Do we still need this if we have the new test? Does this test anything not tested below?

rinaok replied:
this test only tests the corners.

YonatanKra commented on libs/components/src/lib/popup/ui.test.ts:
```diff
@@ -0,0 +1,105 @@
+		);
+});
+
+test('should show all positionings', async ({ page }: { page: Page }) => {
```
![image](https://user-images.githubusercontent.com/6459899/161904099-b45c5fe0-653e-4f03-bbbe-a3c73568e95e.png)
